api/app/Service/UserActionService.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace App\Service;
+
+use App\Http\Controllers\Admin\AdminController;
+use App\Mail\UserBlockedEmail;
+use App\Mail\UserUnblockedEmail;
+use App\Models\User;
+use Illuminate\Support\Facades\Mail;
+
+class UserActionService
+{
+    public function block(User $user, string $reason, int $moderatorId): User
+    {
+        $meta = $user->meta ?? [];
+        $meta['block_reason'] = $reason;
+        $meta['blocked_at'] = now();
+        $meta['blocked_by'] = $moderatorId;
+        $user->meta = $meta;
+        $user->save();
+
+        $user->forms()->update(['visibility' => 'draft']);
+
+        AdminController::log('User blocked', [
+            'user_id' => $user->id,
+            'reason' => $reason,
+            'moderator_id' => $moderatorId,
+        ]);
+
+        Mail::to($user)->send(new UserBlockedEmail($user, $reason));
+
+        return $user->fresh();
+    }
+
+    public function unblock(User $user, string $reason, int $moderatorId): User
+    {
+        $meta = $user->meta ?? [];
+        unset($meta['block_reason']);
```

Comment:
unblock meta should be associated with the block, need to improve data model,

Maybe blocking_history: [
  [
      'reason' => ,
       'blocked_at' => ,
       'blocked_by' => ,
       'unblock_reason' => , // default null
        'unblocked_at' => , // default null
                'unblocked_by' => , // default null
  ]
]

Reply:
Let's create a method to handle these meta changes properly, with one liners indtead of doing $meta[...] too many times